subsys/llext/llext_load.c

```diff
@@ -737,8 +738,8 @@ int do_llext_load(struct llext_loader *ldr, struct llext *ext,
 
 	if (ldr->sect_hdrs_on_heap) {
 		llext_free(ldr->sect_hdrs);
+		ldr->sect_hdrs = NULL;
 	}
-	ldr->sect_hdrs = NULL;
 
```

Comment:
@pillo79 yep, agree, I don't like that either and I expected objections against it. In your first option - you mean `llext_unload()`, right? But currently it has no access to the loader. So we'd need to either modify the function's signature by adding it as the second argument, or to add a pointer to the loader to `struct llext`?

Reply:
The loader struct can be dropped as soon as `llext_load` completes, so I would not create such a constraint. I was instead thinking about fully moving the field from `struct llext_loader` to `struct llext`. That way it is unambiguous that it's a permanent feature. 

The pointer could be freed at the end of `llext_load` unless a load_param field like `keep_section_info` (or whatever) was set, in which case it would be kept until `llext_unload`. 
What do you think?

Reply:
@pillo79 in principle that sounds good, would work fine for me, but with that header array you anyway then have to scan it, you probably don't have section index to jump directly to it. And we do already have an API function `llext_find_section()` that scans that array. Its usefulness at least for SOF is currently limited by it only returning an offset, not a complete header. Ideally I would modify that function to also return an entire section header in a user-supplied parameter, but it's an official API (LLEXT is still experimental). We could add a new function for that and potentially deprecate / migrate over time.
Also, if we do move the pointer to `struct llext` maybe just keep it until freeing without a parameter - it isn't that much memory

Reply:
Great idea! Something along the lines of:
```
int llext_get_section_info(void **start, size_t *length, struct elf_shdr **shdr)
```
where each parameter is optional? 
Or even better, I guess we can have a `struct llext_section_info` with all those fields. It will make it easier to extend in the future without breaking compatibility. 
WDYT?

Reply:
@pillo79 I think that function needs a `name` parameter ;-) And `llext` of course. But `length` is already in the section header? As for `start` - you mean its address in RAM? But that's only available if `peek()` is implemented? In SOF we do
```
offset = llext_find_section(...);
addr = llext_peek(ldr, offset);
```
So maybe only `llext`, `name` and `shdr` as parameters?

Reply:
I like it

Reply:
Sorry, I had a different use case in mind - for _loaded_ sections. LGTM as well!

Reply:
> I like it

@teburd hm, but I've just realised, that it cannot work (without large-scale changes to LLEXT). All access to section header tables in LLEXT is currently performed via the loader. IIUC originally the loader was assumed to be used only for initial linking / relocation, only while `llext_load()` is running. Then we've relaxed that limitation by allowing the user to call `llext_seek()`, `llext_peek()`, `llext_read()` and potentially `llext_find_section()` - all with the loader as the first argument. But still we free all the dynamically allocated loader assets when exiting `llext_load()`. Maybe we indeed should extend the loader life-cycle and make an explicit `loader_free()`?

Reply:
This gets a bit awkward at some point as we're allocating in llext_load() but then I guess we're needing information past the load call. What lifetime are you looking for in the loader? Why?

Reply:
> we've relaxed that limitation by allowing the user to call `llext_seek()`, `llext_peek()`, `llext_read()` 

@lyakh These are used by LLEXT for ELF loading and they are [explicitly undocumented](https://github.com/zephyrproject-rtos/zephyr/blob/main/include/zephyr/llext/loader.h#L118). TBH I don't think an end user should use them :slightly_smiling_face: 

> we free all the dynamically allocated loader assets when exiting `llext_load()`

That should be fine, if we move `sect_hdrs` to `struct llext` as discussed. This is definitely a big commit, but should be automated with a `sed` command :slightly_smiling_face: